In [2]:
import pandas as pd
import numpy as np
import math
from sklearn.model_selection import train_test_split

In [3]:
#dataset = '../Dataset/Car/car'
#dataset = '../Dataset/Balance-sheet/train.txt'
#dataset = '../Dataset/Hayes-Roth/Hayes-Roth'
#dataset = '../Dataset/Heart-Data/train.txt'
dataset = '../Dataset/TicTacToe/tic tac toe'
DATASET_SPLIT_RATIO = .1
np.random.seed(2)

In [4]:
dataframe = pd.read_csv(dataset, header=None)
train, test = train_test_split(dataframe, test_size=DATASET_SPLIT_RATIO)
list_of_features = list(dataframe)
list_of_labels = np.unique(train[list_of_features[-1]], return_counts=True)[0]
list_of_labels_count = np.unique(train[list_of_features[-1]], return_counts=True)[1]

In [5]:
train.head()
print('Number of training samples:', len(train))
print('Number of test samples:', len(test))
print('Labels:', list_of_labels)

Number of training samples: 862
Number of test samples: 96
Labels: ['negative' 'positive']


In [6]:
def predict(data):
    mx = -1000
    mx_idx = -1
    for label_itr in range(len(list_of_labels)):
        probability = np.log(list_of_labels_count[label_itr]/len(train))
        for i in range(len(list_of_features) - 1):
            match = train[(train[i] == data[i]) & (train[list_of_features[-1]] == list_of_labels[label_itr])]
            probability += np.log((len(match)/list_of_labels_count[label_itr]))
        if mx_idx == -1 or probability > mx:
            mx = probability
            mx_idx = label_itr
    return list_of_labels[mx_idx]

In [7]:
right_predictions = 0
for i in range(len(test)):
    if predict(test[i:i+1].values[0]) == test[i:i+1][list(test)[-1]].values[0]:
        right_predictions+=1

In [8]:
print('Accuacy:', str(  round(right_predictions/len(test) * 100, 2)) + '%' )

Accuacy: 68.75%


Car: 83.82%
Balance Sheet: 90.48%
Hayes-Roth: 64.29%
Heart Data: 84.21%
TicTacToe: 68.75%

In [9]:
test[0:1]

0  1  2  3  4  5  6  7  8         9
665  x  x  b  o  o  o  x  o  x  negative

In [39]:
unique_values = []
for i in range(len(list_of_features) - 1):
    unique_values.append(len(np.unique(dataframe[i])))

In [40]:
unique_values

[0, 0, 0, 0, 0, 0, 0, 0, 0]

In [41]:
list_of_features

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

3